drivers/serial/serial_shell.c

```diff
@@ -0,0 +1,163 @@
+/* Copyright (C) 2024 Bang & Olufsen A/S, Denmark
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/drivers/uart.h>
+#include <zephyr/shell/shell.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(serial_shell, CONFIG_UART_LOG_LEVEL);
+
+/* both are 1 as they're not used in the same context */
+#define ARGV_DEV     1
+#define ARGV_TX_DATA 1
+
+K_SEM_DEFINE(busy_sem, 1, 1);
+
+const struct device *enabled_device;
+static char rx_buffer[CONFIG_SERIAL_SHELL_RX_BUFFER_SIZE];
+static char tx_buffer[CONFIG_SERIAL_SHELL_TX_BUFFER_SIZE];
+static uint8_t tx_write_amount;
+static uint8_t tx_buffer_index;
+
+static void uart_isr(const struct device *dev, void *user_data)
+{
+	uart_irq_update(dev);
+
+	if (uart_irq_rx_ready(dev)) {
+		uint8_t c;
+		uint8_t rx_buffer_index = 0;
+
+		while (1) {
+			if (rx_buffer_index >= CONFIG_SERIAL_SHELL_RX_BUFFER_SIZE) {
+				LOG_INF("RX[%u]: %.*s", CONFIG_SERIAL_SHELL_RX_BUFFER_SIZE,
+					CONFIG_SERIAL_SHELL_RX_BUFFER_SIZE, rx_buffer);
```

Comment:
Simply logging isn't what you want probably, it's possible to have logging and shell on different backends, so you will not see any of these prints.

Add a `read` command to get the contents of the `rx_buffer`? Also a [ring buffer](https://docs.zephyrproject.org/latest/kernel/data_structures/ring_buffers.html) might be useful here.

Reply:
> Simply logging isn't what you want probably, it's possible to have logging and shell on different backends, so you will not see any of these prints.

So instead store the shell instance and print from that?

> Add a read command to get the contents of the rx_buffer? Also a [ring buffer](https://docs.zephyrproject.org/latest/kernel/data_structures/ring_buffers.html) might be useful here.

Sure, can do that :+1: Will look into using the `ring_buffer`.

Reply:
I think it's a bit more complex to be honest, what about binary protocols? Do you need to block on reads with a timeout? What's the delimiter when reading? ...

Reply:
> I think it's a bit more complex to be honest, what about binary protocols?

Currently it's limited to character protocols. 
Can add support for binary protocols as well. 

Would be useful to also support runtime configurations (if enabled). Eg:

```
serial set_config <device> <baud/parity/stopbit/databits/flowcontrol> (dicts)
serial set_config <device> <baud/parity/stopbit/databits/flowcontrol> (dicts)
serial get_config <device> -> print <baud/parity/stopbit/databits/flowcontrol>
serial set_protocol <device> <character/binary> (dicts)
serial get_protocol <device> -> print <character/binary>
```

I didn't initially add this feature as we were able to verify hardware with the current handles. I'm happy to extend it if needed.

> Do you need to block on reads with a timeout? What's the delimiter when reading? ...

Synchronization should happen on driver level. 
We print the RX buffer when the driver is done receiving. If the buffer is full it's dumped prematurely.
Currently no delimiter is added but as the buffer is dumped as characters character based protocols should be supported.